# Wheat Yield Data

Ella: 
Start cleaning, beging initial preprocessing
- standardising names
- filter for wheat crop
- remove duplicates
- handle missing values

## Data Loading

In [52]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [38]:
# Load dataset 
df = pd.read_csv('Production_Crops_Livestock_E_All_Data.csv')

C:\Users\Ella\AppData\Local\Temp\ipykernel_15484\4035631550.py:2: DtypeWarning: Columns (11,14,17,20,23,26,29,32,35,38,41,44,47,50,53,56,59,62,65,68,71,74,77,80,83,86,89,92,95,98,101,104,107,110,113,116,119,122,125,128,131,134,137,140,143,146,149,152,155,158,161,164,167,170,173,176,179,182,185,188,191,194,197) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Production_Crops_Livestock_E_All_Data.csv')


#### Initial Date Inspection

In [ ]:
# Display first few rows
df.head()

Area Code Area Code (M49)         Area  Item Code Item Code (CPC)  \
0          2            '004  Afghanistan        221          '01371   
1          2            '004  Afghanistan        221          '01371   
2          2            '004  Afghanistan        221          '01371   
3          2            '004  Afghanistan        711          '01654   
4          2            '004  Afghanistan        711          '01654   

                                                Item  Element Code  \
0                                  Almonds, in shell          5312   
1                                  Almonds, in shell          5412   
2                                  Almonds, in shell          5510   
3  Anise, badian, coriander, cumin, caraway, fenn...          5312   
4  Anise, badian, coriander, cumin, caraway, fenn...          5412   

          Element   Unit  Y1961  ... Y2020N    Y2021  Y2021F Y2021N    Y2022  \
0  Area harvested     ha    0.0  ...    NaN  36862.0       A    NaN  36462.0   
1           Yield  kg/ha    NaN  ...    NaN   1743.2       A    NaN   1742.0   
2      Production      t    0.0  ...    NaN  64256.0       A    NaN  63515.0   
3  Area harvested     ha    NaN  ...    NaN  25357.0       E    NaN  25403.0   
4           Yield  kg/ha    NaN  ...    NaN    705.0       E    NaN    704.0   

   Y2022F Y2022N    Y2023  Y2023F Y2023N  
0       A    NaN  37000.0       A    NaN  
1       A    NaN   1810.8       A    NaN  
2       A    NaN  67000.0       A    NaN  
3       E    NaN  25439.0       E    NaN  
4       E    NaN    704.0       E    NaN  

[5 rows x 198 columns]

In [ ]:
# Check dataset structure and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79711 entries, 0 to 79710
Columns: 198 entries, Area Code to Y2023N
dtypes: float64(63), int64(3), object(132)
memory usage: 120.4+ MB


In [ ]:
# View summary statistics
df.describe()

Area Code     Item Code  Element Code         Y1961         Y1962  \
count  79711.000000  79711.000000  79711.000000  5.737000e+04  5.737700e+04   
mean    1313.302568    865.879251   5408.833310  1.514965e+06  1.549240e+06   
std     2194.895639   1461.515725    101.505636  1.586835e+07  1.615213e+07   
min        1.000000     15.000000   5111.000000  0.000000e+00  0.000000e+00   
25%       89.000000    358.000000   5312.000000  8.362500e+02  8.625000e+02   
50%      169.000000    587.000000   5412.000000  7.500000e+03  7.700000e+03   
75%      272.000000   1037.000000   5510.000000  6.655310e+04  6.874600e+04   
max     5817.000000  17530.000000   5513.000000  1.342210e+09  1.360092e+09   

              Y1963         Y1964         Y1965         Y1966         Y1967  \
count  5.737600e+04  5.739100e+04  5.738600e+04  5.744700e+04  5.745100e+04   
mean   1.582932e+06  1.630178e+06  1.663916e+06  1.704733e+06  1.749159e+06   
std    1.645200e+07  1.694783e+07  1.731071e+07  1.775626e+07  1.817071e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    8.929750e+02  9.100000e+02  9.343750e+02  9.500000e+02  9.885000e+02   
50%    7.981640e+03  8.000000e+03  8.130950e+03  8.400000e+03  8.647000e+03   
75%    7.060525e+04  7.305150e+04  7.500000e+04  7.721212e+04  7.975700e+04   
max    1.368767e+09  1.381032e+09  1.397533e+09  1.408673e+09  1.428836e+09   

       ...         Y2014         Y2015         Y2016         Y2017  \
count  ...  7.103000e+04  7.109600e+04  7.118300e+04  7.141000e+04   
mean   ...  3.156620e+06  3.184441e+06  3.226641e+06  3.273442e+06   
std    ...  3.576517e+07  3.611638e+07  3.656186e+07  3.692842e+07   
min    ...  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  1.524205e+03  1.543307e+03  1.560000e+03  1.557030e+03   
50%    ...  1.207270e+04  1.223683e+04  1.234740e+04  1.237300e+04   
75%    ...  1.211520e+05  1.225511e+05  1.234405e+05  1.241075e+05   
max    ...  2.818170e+09  2.837867e+09  2.912244e+09  2.961412e+09   

              Y2018         Y2019         Y2020         Y2021         Y2022  \
count  7.072900e+04  7.082200e+04  7.093500e+04  7.099100e+04  7.112000e+04   
mean   3.329584e+06  3.345364e+06  3.384600e+06  3.464475e+06  3.504832e+06   
std    3.735102e+07  3.722087e+07  3.759316e+07  3.870242e+07  3.929182e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    1.442300e+03  1.433050e+03  1.432300e+03  1.437730e+03  1.441750e+03   
50%    1.222000e+04  1.225000e+04  1.218570e+04  1.214760e+04  1.220000e+04   
75%    1.255072e+05  1.269745e+05  1.263125e+05  1.270417e+05  1.278680e+05   
max    2.901857e+09  2.964207e+09  3.008965e+09  3.070651e+09  3.073689e+09   

              Y2023  
count  6.614500e+04  
mean   3.752891e+06  
std    4.156594e+07  
min    0.000000e+00  
25%    1.389000e+03  
50%    1.174500e+04  
75%    1.227800e+05  
max    3.134294e+09  

[8 rows x 66 columns]

Observations
- 79711 rows, 198 colu,ns 
- Data types floats for numeric columns (yields, area, production), object for strings and meta data
- Some columns have N, F this could be forecasted data

Missing values 
- Some years have NAN for certain crops in some countries 
- This will need to be handled in data cleaning

Range and Scale
- Yield is in hundreds to thousands
- Area harvested and prduction can go up to billions
- There are outliers

## Data Cleaning

In [53]:
# Standardise column names
# Start with country names
# Remove quotes and extra spaces for consistency
df['Area'] = df['Area'].str.replace("'", "").str.strip()

# Clean crop names by removing extra spaces 
df['Item'] = df['Item'].str.strip()

df[['Area', 'Item']].head()

Area                                               Item
0  Afghanistan                                  Almonds, in shell
1  Afghanistan                                  Almonds, in shell
2  Afghanistan                                  Almonds, in shell
3  Afghanistan  Anise, badian, coriander, cumin, caraway, fenn...
4  Afghanistan  Anise, badian, coriander, cumin, caraway, fenn...

In [ ]:
# Filter to Wheat data only

# Create a dataframe for wheat production only
wheat_df = df[df['Item'] == 'Wheat']

# Check first few rows
wheat_df.head()

Area Code Area Code (M49)         Area  Item Code Item Code (CPC)   Item  \
196          2            '004  Afghanistan         15           '0111  Wheat   
197          2            '004  Afghanistan         15           '0111  Wheat   
198          2            '004  Afghanistan         15           '0111  Wheat   
539          3            '008      Albania         15           '0111  Wheat   
540          3            '008      Albania         15           '0111  Wheat   

     Element Code         Element   Unit      Y1961  ... Y2020N      Y2021  \
196          5312  Area harvested     ha  2230000.0  ...    NaN  2046564.0   
197          5412           Yield  kg/ha     1022.0  ...    NaN     1963.1   
198          5510      Production      t  2279000.0  ...    NaN  4017657.0   
539          5312  Area harvested     ha   126500.0  ...    NaN    54514.0   
540          5412           Yield  kg/ha      773.1  ...    NaN     4130.5   

     Y2021F Y2021N      Y2022  Y2022F Y2022N      Y2023  Y2023F  \
196       A    NaN  1859339.0       A    NaN  2003070.0       I   
197       A    NaN     2045.3       A    NaN     2146.7       E   
198       A    NaN  3802895.0       A    NaN  4300000.0       X   
539       A    NaN    54267.0       A    NaN    57620.0       A   
540       A    NaN     4296.3       A    NaN     4053.9       A   

                Y2023N  
196                NaN  
197                NaN  
198  Unofficial figure  
539                NaN  
540                NaN  

[5 rows x 198 columns]

In [55]:
# Filter for rows where the element is yield 
wheat_df = wheat_df[wheat_df['Element'] == 'Yield']

wheat_df.head

<bound method NDFrame.head of        Area Code Area Code (M49)                                     Area  \
197            2            '004                              Afghanistan   
540            3            '008                                  Albania   
889            4            '012                                  Algeria   
1125           7            '024                                   Angola   
1720           9            '032                                Argentina   
...          ...             ...                                      ...   
77152       5801            '199                Least Developed Countries   
77806       5802            '432         Land Locked Developing Countries   
78299       5803            '722           Small Island Developing States   
78958       5815            '901        Low Income Food Deficit Countries   
79640       5817            '902  Net Food Importing Developing Countries   

       Item Code Item Code (CPC)   Item  Elem

In [ ]:
# Handle missing values 
# Check missing values
print("Missing values per column:")
wheat_df.isnull().sum()

Missing values per column:


Area Code            0
Area Code (M49)      0
Area                 0
Item Code            0
Item Code (CPC)      0
                  ... 
Y2022F              10
Y2022N             164
Y2023               10
Y2023F              10
Y2023N             164
Length: 198, dtype: int64

In [ ]:
# Remove flag and note columns (ending in F or N)
# These aren't needed for yield analysis
year_cols_clean = [col for col in wheat_df.columns 
                   if col.startswith('Y') and not col.endswith(('F', 'N'))]

# Keep only Area, Item, Element and clean year columns
keep_cols = ['Area', 'Item', 'Element'] + year_cols_clean
wheat_df_clean = wheat_df[keep_cols]

print(f"Columns after removing flags: {wheat_df_clean.shape[1]}")

Columns after removing flags: 66


In [ ]:
# Remove rows where ALL year values are missing
# (Keep rows with at least some data)
year_cols_clean = [col for col in wheat_df_clean.columns if col.startswith('Y')]
wheat_df_clean = wheat_df_clean.dropna(subset=year_cols_clean, how='all')

print(f"Rows after removing completely empty records: {len(wheat_df_clean)}")

Rows after removing completely empty records: 164


In [ ]:
# For remaining missing values, leave them as NaN
# Not all countries grow all crops in all years
print("Final missing value count:")
wheat_df_clean.isnull().sum()

Final missing value count:


Area        0
Item        0
Element     0
Y1961      40
Y1962      40
           ..
Y2019       9
Y2020       9
Y2021       9
Y2022      10
Y2023      10
Length: 66, dtype: int64

In [59]:
# Check final cleaned dataset shape
print(f"Final cleaned dataset shape: {wheat_df_clean.shape}")

Final cleaned dataset shape: (164, 66)


In [60]:
# Drop duplicate rows if any 
wheat_df = wheat_df.drop_duplicates()
wheat_df.shape

(164, 198)